src/osd/PrimaryLogPG.h

```diff
@@ -1623,17 +1623,30 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
     struct ReservationCB : public Context {
       PrimaryLogPGRef pg;
       bool canceled;
-      ReservationCB(PrimaryLogPG *pg) : pg(pg), canceled(false) {}
+      Mutex lock;
+      ReservationCB(PrimaryLogPG *pg) : pg(pg), canceled(false),
+					lock("ReservationCB", false, false) {}
       void finish(int) override {
-	pg->lock();
-	if (!canceled)
-	  pg->snap_trimmer_machine.process_event(SnapTrimReserved());
-	pg->unlock();
+	PrimaryLogPGRef pgp;
+	bool is_canceled = false;
+	{
+	  Mutex::Locker l(lock);
+	  pgp = pg;
+	  is_canceled = canceled;
+	}
+	if (!pgp)
+	  return;
+	pgp->lock();
+	if (!is_canceled)
+	  pgp->snap_trimmer_machine.process_event(SnapTrimReserved());
+	pgp->unlock();
       }
       void cancel() {
 	assert(pg->is_locked());
+	Mutex::Locker l(lock);
 	assert(!canceled);
 	canceled = true;
+	pg.reset();
```

Comment:
OSD::shutdown() loops through and locks the PGs twice. The first time it calls PrimaryPG::on_shutdown(), which invokes this cancel.
The second time, after doing some other work, is when we need the remaining PGRefs to be gone. So we're okay if the finish() function is actually racing live, since it will get the PG lock and drop the ref before we need it to be gone.